### PR TITLE
Parse module headers

### LINKED ISSUE #505: Avoid parsing sources for dependencies
Now that ghcide is using interface files, it would be desirable if ghcide parsed only the module headers to spy out the imports and skipped the rest of the file, just like GHC does.

### PR DESCRIPTION
**Pull request by pepeiborra**
Instead of full modules, ghcide now parses only the module headers and avoids storing unnecessary ASTs in the Shake graph where possible. This happens in a couple of places:

* Locating the imports of a module. Now this uses a `ModSummary` instead of a `ParsedModule`
* For computing all the completions in a module. Now this uses a `ModIface` instead, i.e. a `.hi` file. 
* For computing source span infos, the parsed modules are only used when building against `ghc-lib`.  Otherwise, `ModIface` is used instead. 
* Interface files created by ghcide now embed Haddocks

Altogether, ghcide only parses full modules for files of interest now, or for dependencies when an up-to-date `.hi` file is not available. The main benefits are 
1. shorter start-up times 
2. lower memory usage. 

Joint work with @lazamar.

Fixes #505

### CODE REVIEW

File: src/Development/IDE/Core/Compile.hs [hunk 1]
```diff
@@ -447,8 +449,46 @@ getModSummaryFromBuffer fp contents dflags parsed = do
     where
       sourceType = if "-boot" `isSuffixOf` takeExtension fp then HsBootFile else HsSrcFile
 
-
--- | Given a buffer, flags, file path and module summary, produce a
+-- | Given a buffer, env and filepath, produce a module summary by parsing only the imports.
+--   Runs preprocessors as needed.
+getModSummaryFromImports :: FilePath -> Maybe SB.StringBuffer -> HscEnv -> ExceptT [FileDiagnostic] IO ModSummary
```
Comment (mpickering) [inline src/Development/IDE/Core/Compile.hs]: As far as I can see (from quite a brief inspection) the only information ever used from this `HscEnv` is the `DynFlags`, so it should be safe to pass in ` DynFlags` here and then pass in a faked `HscEnv` to `preprocess`.

File: src/Development/IDE/Core/Compile.hs [hunk 2]
```diff
@@ -447,8 +449,46 @@ getModSummaryFromBuffer fp contents dflags parsed = do
...
+getModSummaryFromImports fp contents hsc_env = do
+    (contents, dflags) <- ExceptT $ evalGhcEnv hsc_env $ runExceptT $ preprocessor fp contents
+    (srcImports, textualImports, L _ moduleName) <-
+        ExceptT $ first (diagFromErrMsgs "parser" dflags) <$> GHC.getHeaderImports dflags contents fp fp
+    liftIO $ evaluate $ rnf srcImports
+    liftIO $ evaluate $ rnf textualImports
```
Comment (mpickering) [inline src/Development/IDE/Core/Compile.hs]: Comment about why this is necessary.

File: src/Development/IDE/Core/Compile.hs [hunk 3]
```diff
@@ -447,8 +449,46 @@ getModSummaryFromBuffer fp contents dflags parsed = do
...
+    modLoc <- liftIO $ mkHomeModLocation dflags moduleName fp
+    required_by_imports <- liftIO $ implicitRequirements hsc_env textualImports
```
Comment (mpickering) [inline src/Development/IDE/Core/Compile.hs]: Looks like this is copied from something to do with backpack, do we support backpack projects?

`implicitRequirements` calls `findImportedModule` which is probably desirable to avoid here.

File: src/Development/IDE/Core/Compile.hs [hunk 4]
```diff
@@ -432,7 +434,7 @@ getModSummaryFromBuffer fp contents dflags parsed = do
     , ms_textual_imps = [imp | (False, imp) <- imports]
     , ms_hspp_file    = fp
     , ms_hspp_opts    = dflags
-    , ms_hspp_buf     = Just contents
+    , ms_hspp_buf     = Nothing
```
Comment (mpickering) [inline src/Development/IDE/Core/Compile.hs]: Was this field not used before?

File: src/Development/IDE/Core/Rules.hs [hunk 5]
```diff
@@ -342,22 +347,27 @@ getSpanInfoRule :: Rules ()
 getSpanInfoRule =
     define $ \GetSpanInfo file -> do
         tc <- use_ TypeCheck file
+        packageState <- hscEnv <$> use_ GhcSession file
         deps <- maybe (TransitiveDependencies [] [] []) fst <$> useWithStale GetDependencies file
         let tdeps = transitiveModuleDeps deps
+#ifdef GHC_LIB
         parsedDeps <- uses_ GetParsedModule tdeps
+#else
+        let parsedDeps = []
```
Comment (mpickering) [inline src/Development/IDE/Core/Rules.hs]: How does this work in the GHC_LIB case? I don't understand why the CPP is here at all actually.

File: src/Development/IDE/Core/Rules.hs [hunk 6]
```diff
@@ -526,11 +538,22 @@ getModIfaceRule = define $ \GetModIface f -> do
...
-            let !res = extract tmr
-            return (diags, res)
+            -- the interface file does not exist or is out of date.
+            -- Invoke typechecking directly to update it without incurring a dependency
+            -- on the parsed module and the typecheck rules
+            hsc <- hscEnv <$> use_ GhcSession f
+            opt <- getIdeOptions
+            (_, contents) <- getFileContents f
+            -- Embed --haddocks in the interface file
+            hsc <- pure hsc{hsc_dflags = gopt_set (hsc_dflags hsc) Opt_Haddock}
```
Comment (mpickering) [inline src/Development/IDE/Core/Rules.hs]: Why do we set this here and not where we set all the other options? Also seems unrelated to the patch as a whole.

File: src/Development/IDE/Plugin/Completions.hs [hunk 7]
```diff
@@ -22,17 +22,24 @@ import Development.IDE.Core.RuleTypes
...
 
 produceCompletions :: Rules ()
 produceCompletions =
     define $ \ProduceCompletions file -> do
+#ifdef GHC_LIB
         deps <- maybe (TransitiveDependencies [] [] []) fst <$> useWithStale GetDependencies file
         parsedDeps <- mapMaybe (fmap fst) <$> usesWithStale GetParsedModule (transitiveModuleDeps deps)
+#else
+        let parsedDeps = []
+#endif
```
Comment (mpickering) [inline src/Development/IDE/Plugin/Completions.hs]: This is a bit confusing to read. It just seems like this feature won't work with GHC_LIB?

File: src/Development/IDE/Core/Compile.hs (hunk 1 above)
Comment (pepeiborra) [reply src/Development/IDE/Core/Compile.hs]: The `HscEnv` is also required for `implicitRequirements`, which is part of the BackPack logic in GHC. Unless BackPack support is a priority, which probably requires a fair amount of work, I think it's safe to drop the call to `implicitRequirements` and generalize the types to lift the requirement of an `HscEnv`

File: src/Development/IDE/Core/Compile.hs (hunk 3 above)
Comment (pepeiborra) [reply src/Development/IDE/Core/Compile.hs]: ack

File: src/Development/IDE/Core/Compile.hs (hunk 4 above)
Comment (pepeiborra) [reply src/Development/IDE/Core/Compile.hs]: Everything works with `Nothing` as far as I can tell. Git grepping in the ghc repo for the field name doesn't show any relevant usage. It really seems fine to drop it

File: src/Development/IDE/Core/Rules.hs (hunk 5 above)
Comment (pepeiborra) [reply src/Development/IDE/Core/Rules.hs]: The need for this comes from `Development.IDE.Spans.Documentation`. I'll add some comments

File: src/Development/IDE/Core/Rules.hs (hunk 6 above)
Comment (pepeiborra) [reply src/Development/IDE/Core/Rules.hs]: We want to embed haddocks in the interface files, so we set `Opt_Haddock` on the set of flags that we use for generating them. It's pretty crucial for the patch, without this we cannot skip parsing modules for `getSpanInfos` and `produceCompletions`.

`Opt_Haddock` is not desired globally, see https://github.com/digital-asset/ghcide/pull/350#discussion_r370878197

File: src/Development/IDE/Plugin/Completions.hs (hunk 7 above)
Comment (pepeiborra) [reply src/Development/IDE/Plugin/Completions.hs]: Well, since apparently there is an issue that prevents GHC_LIB from reading Haddocks from interface files, it is not possible to avoid parsing sources when Haddocks/completions are required.

**Comment by mpickering** [comment]
An annoying problem I immediately ran into, if a module fails to parse because of the `Opt_Haddock` flag then you get an error diagnostic. As soon as you open this module however, the diagnostic goes away because the `Opt_Haddock` flag is no longer enabled!

**Comment by mpickering** [comment]
I am also getting a lot of traces.

```
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/Match/Constructor.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/Binds.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/PmCheck.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/ListComp.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcSimplify.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/PmCheck/Ppr.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcInteract.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcCanonical.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/Expr.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/Binds.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcInteract.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/Match.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/ListComp.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/Match/Constructor.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/PmCheck/Ppr.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/Arrows.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/PmCheck.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/PmCheck/Oracle.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcSimplify.hs
Released stringbuffer for (preprocessor) /home/matt/core-typedth/compiler/GHC/Iface/Load.hs-boot
Released stringbuffer for (preprocessor) /home/matt/core-typedth/compiler/GHC/CoreToIface.hs-boot
Released stringbuffer for (preprocessor) /home/matt/core-typedth/compiler/GHC/Driver/Plugins.hs-boot
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcRnTypes.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcMType.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Driver/Plugins.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/FamInst.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Rename/Env.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Core/Make.hs
Released stringbuffer for (preprocessor) /home/matt/core-typedth/compiler/GHC/Driver/Pipeline/Monad.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Core/Unfold.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/Inst.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Core/Op/ConstantFold.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/basicTypes/MkId.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Iface/Load.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcFlatten.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Iface/Rename.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Core/Lint.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/CoreToIface.hs
Released stringbuffer for (preprocessor) /home/matt/core-typedth/compiler/GHC/HsToCore/Monad.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Iface/Binary.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/Utils.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/GuardedRHSs.hs
Released stringbuffer for (preprocessor) /home/matt/core-typedth/compiler/typecheck/TcHoleFitTypes.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Iface/Env.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcHsSyn.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/prelude/PrelInfo.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/iface/BuildTyCl.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/DsMetaTc.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcUnify.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Types/Name/Shape.hs
Released stringbuffer for (preprocessor) /home/matt/core-typedth/compiler/GHC/HsToCore/Binds.hs-boot
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Core/SimpleOpt.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Core/Rules.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcEnv.hs
Released stringbuffer for (preprocessor) /home/matt/core-typedth/compiler/GHC/Rename/Unbound.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/ClsInst.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/Match/Literal.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Core/Op/Monad.hs
Released stringbuffer for (preprocessor) /home/matt/core-typedth/compiler/GHC/Rename/Utils.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/HsToCore/Quote.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/GHC/Driver/Finder.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcTypeable.hs
Released stringbuffer for (preprocessor) /home/matt/core-typedth/compiler/GHC/HsToCore/Expr.hs-boot
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcErrors.hs
Released stringbuffer for (runCpp) /home/matt/core-typedth/compiler/typecheck/TcSMonad.hs
```

**Comment by pepeiborra** [comment]
Ah, my bad, those traces were not intended for this PR. Will fix

**Comment by pepeiborra** [comment]
> An annoying problem I immediately ran into, if a module fails to parse because of the `Opt_Haddock` flag then you get an error diagnostic. As soon as you open this module however, the diagnostic goes away because the `Opt_Haddock` flag is no longer enabled!

Yup, that’s going to be annoying. First workaround that comes to mind is parse again without `Opt_Haddock` in case of error, not pretty though

**Comment by pepeiborra** [comment]
> > An annoying problem I immediately ran into, if a module fails to parse because of the `Opt_Haddock` flag then you get an error diagnostic. As soon as you open this module however, the diagnostic goes away because the `Opt_Haddock` flag is no longer enabled!
> 
> Yup, that’s going to be annoying. First workaround that comes to mind is parse again without `Opt_Haddock` in case of error, not pretty though. @cocreature @ndmitchell any other suggestions?

I have gone with a different solution: parse files of interest twice, with and without `Opt_Haddock`, to collect Haddock parse errors consistently.

File: src/Development/IDE/Core/Compile.hs [hunk 8]
```diff
@@ -447,8 +449,51 @@ getModSummaryFromBuffer fp contents dflags parsed = do
...
+        summary =
+            ModSummary
+                { ms_mod          = mod
+#if MIN_GHC_API_VERSION(8,8,0)
+                , ms_hie_date     = Nothing
+#endif
+                , ms_hs_date      = error "Rules should not depend on ms_hs_date"
+        -- When we are working with a virtual file we do not have a file date.
+        -- To avoid silent issues where something is not processed because the date
+        -- has not changed, we make sure that things blow up if they depend on the
```
Comment (cocreature) [inline src/Development/IDE/Core/Compile.hs:481]:
```suggestion
        -- has not changed, we make sure that things blow up if they depend on the date.
```
